### ISSUE TITLE
Filtering expressions called before dataChange Event Fired = Trouble in Paradise

### ISSUE THREAD

**Issue by richieb** [Polymer/polymer#1169]
The filtering expressions are called before any dataChange events are fired, which means you can't do this:
```
<div> {{data.firstName | addAgeToName }} </div>

addAgeToName: function (firstName) {
// this.age is undefined
console.log ("Filter function called");
return firstName + "-" + this.age;
},
dataChanged: function () {
console.log ("dataChanged function called");
this.age = this.data.age;
}
```

Console logged this:
```
Filter function called
dataChanged function called
```
So the filter functions are being called first, which means we **cannot** (UPDATED) use the *data* inside the filter functions.
How do we handle this?

**Comment by richieb** [comment]
I should add that I am using this "hacky" way to get around the issue:
```
<div> {{ {firstName: data.firstName, age:data.age} | addAgeToName }} </div>

addAgeToName: function (nameAndAge) {
return nameAndAge.firstName + "-" + nameAndAge.age;
}
```
I would would rather use the saved data, captured during the *dataChanged* event.

**Comment by arthurevans** [comment]
So I think what's happening here is that setting data queues up two actions, updating the bindings and calling the dataChanged callback. Your data changed then updates a piece of data used by the filter, but the binding system doesn't know it's changed (because age isn't in the binding expression). So your hacky expression is actually more correct in the sense that it expresses the dependencies clearly.

If you add "age" as an argument to the filter, you'll see the same thing -- the filter gets called again after the `dataChanged` callback:

```
<div> {{data.firstName | addAgeToName(age) }} </div>
```

There are a couple of other alternatives. I usually use a straight function in place of a filter -- it just feels like it expresses the dependencies more obviously. You could use a computed property. You could even do it all in the binding:

```
 <div> {{firstName + "-" + age }} </div>
```

For what it's worth, I think the 0.8 binding system is moving towards computed properties for this kind of thing, so I'd be inclined to use that if it works.

**Comment by richieb** [comment]
Thanks for the reply, Arthur.

In my case, simply using computed properties wouldn't work because my *addAgeToName* function is actually complex function that doesn't add a name to age. I used *addAgeToName* as a trivial example to illustrate my point.

In my real code, my filter function has to loop through an object and run some calculations on each item, then format the final data it will return. Also, the filter function depends on the the dataChanged event because the dataChanged function takes part of the main data and modify it a bit and save it to a global object, which will be used by other elements.

The bottom line is that I really need to do the following:
1. Manipulate the data (on the *dataChanged* event) and save it to a global obj.
2. For the filter function, I have to make use of that data I saved in to the global object, and do a bit of computation before I return the formatted content from the filter function.

I am not sure if I understand the first part of your response. Did you say why the filter function is called before the dataChanged function?

**Comment by richieb** [comment]
I think I have an update how to make it work, based on your reply. Let me go back and try a couple of things.

**Comment by richieb** [comment]
Okay, I just tested and it works as you have described, **but the filter function is always being called first**, before the dataChanged function. Indeed, the filter function runs again, after the dataChanged function, if one of the dependable data changes.

I still think is a bit wonky.

**Comment by tjsavage** [comment]
Closing this issue due to age and the release of [version 1](https://www.polymer-project.org/1.0/) of Polymer - please feel free to re-open if this is incorrect.
